File: 백준/Gold/20040. 사이클 게임/사이클 게임.js
let fs = require('fs')
let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')
// let input = fs.readFileSync('./EX.txt').toString().trim().split('\n')
const [N, M] = input[0].split(" ").map(x => parseInt(x))

const p = []
for (let i = 0; i < N; i++) {
    p.push(i)
}

function find_set(x) {
    if (p[x] !== x) {
        p[x] = find_set(p[x])
    } 
    return p[x]
}


let cnt = M + 1
function union(x, y, i) {
    let xp = find_set(x)
    let yp = find_set(y)
    if (xp === yp) {
        cnt = Math.min(cnt, i)
        return
    } else if (xp < yp) {
        p[yp] = xp
    } else {
        p[xp] = yp
    }
}

for (let i = 1; i < input.length; i++) {
    const line = input[i].split(" ").map(x => parseInt(x))
    union(line[0], line[1], i)
}

{cnt === M+1 ? console.log(0) : console.log(cnt)}

File: 백준/Gold/20040. 사이클 게임/사이클 게임.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

n, m = map(int, input().split())
p = [int(i) for i in range(n)]

def find_set (x):
    if p[x] != x:
        p[x] = find_set(p[x])
    return p[x]

def union(x, y, i):
    global cnt
    xp = find_set(x)
    yp = find_set(y)

    if xp == yp:
        cnt = i
        print(i+1)
        exit(0)
    elif xp < yp:
        p[yp] = xp
    else:
        p[xp] = yp

cnt = 0
for i in range(m):
    a, b = map(int, input().split())
    union(a, b, i)

print(cnt)